### PR TITLE
feat(core-logic): add core step function alerting logic

### PR DESCRIPTION
**Pull request by benjaminpaige**
## Purpose

This change set adds the core step function logic to compare MMDL and SEA Tool records and send alerts based on their status. 

#### Linked Issues to Close

https://qmacbis.atlassian.net/browse/OY2-21832

## Approach

The compare service was built out to handle records streaming in from the mmdl table, hydrated from the connector service, and initiate step function logic for relevant records. 

the compare step function logic was expanded to handle record logic according to the requirements [here](https://qmacbis.atlassian.net/browse/OY2-20141)

The basic logic of the compare service is as follows:

the wokflowStarter lambda has an mmdl stream which receives all change events to the mmdl table. this function retrieves the mmdl item, uses the getMmdlInfoFromRecord util to determine if the record has been signed and if so, how long ago. If the record has not been signed yet or if its older than 250 days, the function ignores the record. Otherwise it starts the compare state machine with the record id as input. 

The state machine is kicked off with the initStatus function which initializes a record in the status table.
The service then waits for a time determined using stage params. These params use second units and are identical for higher environments but use smaller increments for default branches to allow for a more immediate feedback loop when testing. 

After the initialWait the mmdl record is retrieved from the mmdl dynamo table, we check to see if a seatool record exists for the same id, if one exists we compare the records to see if they were both signed on the same date, if not an email is sent to the recipients defined by the values returned from secrets manager indicating that no match has been found. 

if no seatool record is found with a matching id, an email indicating no record exists is sent in the same manner. 

The record is then updated in the status table to reflect the new data, and if the record is now older than 250 days and is signed, it is considered successful, if not 250 days old it will go into a waiting stage and cycle through the machine again. Eventually the record will be older than 250 days old and be a match or will fail : ( 

Some additional relevant changes
- Logging
  - Add step function log group to collect step function service log 
  - Add Error SNS Topic and collect all sf errors for added visibility
  - Add Log Group and streams which log all emails sent from the alerting service
- Libs
  - putLogEvent was added to the cloudwatch-lib
  - getItem function added to dynamodb-lib which will allow item puts to dynamo tables
  - ses-lib added with sendAlert function to send emails
  - sns-lib added with trackError function used to send errors to the errorsTopic 
  - rollup file added as libs/index.js
- Status service 
  - status dynamo table added used to collect holistic status updates from items as they move through compare sf machine

## Assorted Notes/Considerations

There will still need to be documentation added to the docs site for all of this, this has been scoped out into a [separate ticket](https://qmacbis.atlassian.net/browse/OY2-21833)

we are using secrets manager to store values for email recipients/senders. the path of these values are based on the stage, we will need to add this secrets value for higher environments. If no secret is found for a particular stage, which will be the case for ephemeral environments unless purposefully set, the emails will not be sent, but they will be logged to the respective log stream.

Stages in the compare step function logic were named with the following pattern: 
- Lambdas which perform some action with the data are post-faced with 'Task' ex. SendNotExistTask
- Wait staged with 'Wait' ex. InitialWait
- Choices which determine the next step based on a value of the data provided use 'Choice'. ex. SignedSinceChoice

Proposed step function logic: 
image

Resulting step function logic after this change:
image

### CODE REVIEW

File: src/libs/sns-lib.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  const params = {
+    Message: JSON.stringify(e, null, 2),
+    TopicArn: process.env.errorsTopicArn,
+  };
+  try {
+    const data = await snsClient.send(new PublishCommand(params));
+    return data; // For unit tests.
+  } catch (err) {
+    console.error(
+      "Error traking the error - a meta error",
```
Comment (benjaminpaige) [inline src/libs/sns-lib.js:16]:
```suggestion
      "Error tracking the error - a meta error",
```

File: src/services/alerts/serverless.yml [hunk 2]
```diff
@@ -27,11 +28,21 @@ custom:
 
 resources:
   Resources:
+    SentAlertsGroup:
+      Type: AWS::Logs::LogGroup
+      Properties:
+        RetentionInDays: 365
```
Comment (benjaminpaige) [inline src/services/alerts/serverless.yml:34]: negative to save foreva?

File: src/services/compare/serverless.yml [hunk 3]
```diff
@@ -73,60 +107,188 @@ functions:
...
+          HandleSeatoolRecordExistChoice:
+            Type: Choice
+            Choices:
+              - Variable: $.seatoolExist
+                BooleanEquals: true
+                Next: CompareTask
+              - Variable: $.seatoolExist
+                BooleanEquals: false
+                Next: SignedSinceChoice
+            Default: SignedSinceChoice
```
Comment (benjaminpaige) [inline src/services/compare/serverless.yml:223]: if this will never happen, and is not required, does this need a default?

Comment (benjaminpaige) [reply src/services/compare/serverless.yml:223]: or maybe a default to catch weirdness?

File: src/services/compare/handlers/workflowStarter.js [hunk 4]
```diff
@@ -1,22 +1,49 @@
...
+  - mmdlSigned: boolean
+  - secSinceMmdlSigned?: number */
+  const sigInfo = getMmdlSigInfo(mmdlRecord);
+
+  /* Checking if the mmdl was signed within the last 250 days. */
+  if (sigInfo.mmdlSigned && sigInfo.secSinceMmdlSigned < 21686400) {
+    /* Creating an object that will be passed to the StartExecutionCommand. */
+    const params = {
+      input: JSON.stringify({
+        id: id,
```
Comment (benjaminpaige) [inline src/services/compare/handlers/workflowStarter.js:25]:
```suggestion
        id,
```

File: src/services/compare/serverless.yml (hunk 3 above)
Comment (benjaminpaige) [reply src/services/compare/serverless.yml:223]: default is required.

**Review by JonHolman — APPROVED** [review]

**Pull request merged**